includes/Core/Tags/Tag.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Tags\Tag
+ *
+ * @package   Google\Site_Kit\Core\Tags
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Tags;
+
+use Google\Site_Kit\Core\Guards\Guard_Interface;
+
+/**
+ * Base class for tags.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+abstract class Tag implements Tag_Interface {
+
+	/**
+	 * Module slug.
+	 *
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	protected $slug;
+
+	/**
+	 * Tag ID.
+	 *
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	protected $tag_id;
+
+	/**
+	 * Guards array.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $guards = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $module_slug Module slug.
+	 * @param string $tag_id Tag ID.
+	 */
+	public function __construct( $module_slug, $tag_id ) {
+		$this->module_slug = $module_slug;
+		$this->tag_id      = $tag_id;
+	}
+
+	/**
+	 * Adds a new guard to the guards list.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Guard_Interface $guard A guard instance to add to the guards list.
+	 */
+	public function use_guard( Guard_Interface $guard ) {
+		$this->guards[] = $guard;
+	}
+
+	/**
+	 * Registers tag if it can be activated.
+	 *
+	 * @since n.e.x.t
+	 */
+	final public function register() {
+		foreach ( $this->guards as $guard ) {
+			if ( $guard instanceof Guard_Interface ) {
+				$can_activate = $guard->can_activate();
+				if ( is_wp_error( $can_activate ) || ! $can_activate ) {
+					return;
+				}
+			}
+		}
```

Comment:
I'm not convinced of the "guard" approach in this PR. I think it would make sense if these were truly standalone/agnostic `Tag` classes, but all of these are more `Module_Tag` classes. We could potentially introduce an even lower level in the future, but I think that would be better as a follow-up PR.

I suggest to do the following:
* Stick with the IB for relying on `Module_Tag` classes that include their own logic based on `Module_Settings` instances for whether they should render or not etc.
* Add a follow-up issue for low-level `Tag` classes that are not associated with modules at all and could use more abstract logic like the `Guard` approach. The `Module_Tag` classes could then be migrated to use `Tag` classes internally.